docs/design/virtual-addresses.md

```diff
@@ -0,0 +1,135 @@
+Virtual addresses will resolve the issue of account EVM compatibility and identification and will be responsible for EVM address related logic.
+
+## Goals
+- Add list of evmAddress values known as “Virtual Address” to Hedera accounts which govern the address the EVM observes for a given account transaction
+- Hedera Accounts can add, disable and remove virtual address entries as desired
+- The Hedera network will:
+  - Validate ownership by extracting the public key from the signature and comparing the calculated public address to the `evmAddress` passed in on addition of the virtual address
+  - Maintain an `evmAddress -> accountId` map thereafter
+- Contract accounts may utilize the `evmAddress` to store their address in accordance with `CREATE` and `CREATE2` EVM operations
+- Perform seamless state migration of accounts and contracts so that by default they benefit from the introduction of virtual addresses
+- Restoring HIP 32’s consistency so that account aliases will have only key based values
+
+## Non Goals
+- Support removal of virtual addresses between accounts
+- Support transferring of virtual addresses between accounts
+
+## Architecture
+
+### Account state updates
+- Update `HederaAccount` interface and its implementations (`MerkleAccount` & `OnDiskAccount`) to have the ability to read/write virtual addresses and default virtual address
+- Update `MerkleAccountState` with matching changes as in `HederaAccount`
+- Nonce management
+  - Update account state definition to track nonce value for each virtual address
+  - Nonce updates will be externalized by adding a map `ContractId -> nonce` in message `ContractFunctionResult` that will be populated in the top-level transaction records.
+
+#### Separate EVM address lookup from aliases
+- Introduce a new class for managing `evmAddress -> accountId` in-memory map
+- The `evmAddress -> accountId` will be initially filled during the state migration
+- Implement logic to rebuild the `evmAddress -> accountId` map from state e.g. similar to `AliasManager.rebuildAliasesMap` implementation
+- Update `AliasManager.rebuildAliasesMap` method to work only for EC/ED key bytes aliases 
+
+#### Introduce special system account
+- Create new system account for exposed keys/addresses from Ethereum tools and external chains that will be used to ensure that no such keys/addresses will be allowed 
+- The limitation for accounts to have a maximum limit of virtual addresses will not be applicable to this system account i.e. this account will have unbounded list of virtual addresses 
+
+#### State migration
+- Create new class in `com.hedera.node.app.service.mono.state.migration` that implements the migration logic for the accounts and contracts
+- During migration all `evmAddress` entries that are currently in `AliasManager` map should be moved to the `evmAddress -> accountId` map
```

Comment:
Do we have to move/set the ethereum_nonce

Reply:
Yes, I think we should I will be updating the HIP and will add it there in the migration steps.